File: packages/tapsdk-pc-js/ts/achievement.ts
/**
 * TapTap PC SDK - Achievement functionality
 */

import { native } from './native.js';

/**
 * Achievement API
 *
 * @example
 * ```typescript
 * import { Achievement, EventId } from 'tapsdk-pc';
 *
 * const achievement = Achievement.get();
 *
 * achievement.unlock(1, 'first_win');
 * achievement.increment(2, 'kill_100_enemies', 1);
 * achievement.showAchievements();
 * ```
 */
export class Achievement {
  private readonly _native: ReturnType<typeof native.Achievement.get>;

  private constructor(nativeInstance: ReturnType<typeof native.Achievement.get>) {
    this._native = nativeInstance;
  }

  /**
   * Get the achievement singleton instance.
   *
   * @returns Achievement instance
   * @throws Error if SDK is not initialized
   */
  static get(): Achievement {
    const nativeInstance = native.Achievement.get();
    return new Achievement(nativeInstance);
  }

  /**
   * Unlock an achievement.
   *
   * The result will be delivered via the AchievementUnlock event.
   *
   * @param requestId - A unique ID to identify this request in the callback
   * @param achievementId - The achievement identifier configured in TapTap
   */
  unlock(requestId: number, achievementId: string): void {
    this._native.unlock(requestId, achievementId);
  }

  /**
   * Increment progress for a step-based achievement.
   *
   * The result will be delivered via the AchievementIncrement event.
   *
   * @param requestId - A unique ID to identify this request in the callback
   * @param achievementId - The achievement identifier configured in TapTap
   * @param steps - The number of steps to add
   */
  increment(requestId: number, achievementId: string, steps: number): void {
    this._native.increment(requestId, achievementId, steps);
  }

  /**
   * Open the TapTap achievements page.
   */
  showAchievements(): void {
    this._native.showAchievements();
  }
}

File: crates/tapsdk-pc/src/achievement.rs
//! Achievement functionality

use std::ffi::CString;

use crate::error::{Result, SdkResult, TapSdkError};
use crate::sdk::is_initialized;

/// Achievement API handle.
///
/// Get an instance via `Achievement::get()` after initializing the SDK.
pub struct Achievement {
    handle: *mut tapsdk_pc_sys::ITapAchievement,
}

// The SDK exposes achievements through a process-wide singleton.
unsafe impl Send for Achievement {}
unsafe impl Sync for Achievement {}

impl Achievement {
    /// Get the achievement singleton instance.
    pub fn get() -> Option<Self> {
        if !is_initialized() {
            return None;
        }

        let handle = unsafe { tapsdk_pc_sys::TapAchievement() };

        if handle.is_null() {
            None
        } else {
            Some(Achievement { handle })
        }
    }

    /// Unlock an achievement.
    ///
    /// The result will be delivered via the `AchievementUnlock` event when
    /// calling `TapSdk::run_callbacks()`.
    pub fn unlock(&self, request_id: i64, achievement_id: &str) -> Result<()> {
        let achievement_id_c = CString::new(achievement_id)?;
        let request = tapsdk_pc_sys::TapAchievementUnlockRequest {
            achievement_id: achievement_id_c.as_ptr(),
        };

        let result =
            unsafe { tapsdk_pc_sys::TapAchievement_AsyncUnlock(self.handle, request_id, &request) };

        check_sdk_result(result)
    }

    /// Increment progress for a step-based achievement.
    ///
    /// The result will be delivered via the `AchievementIncrement` event when
    /// calling `TapSdk::run_callbacks()`.
    pub fn increment(&self, request_id: i64, achievement_id: &str, steps: u64) -> Result<()> {
        let achievement_id_c = CString::new(achievement_id)?;
        let request = tapsdk_pc_sys::TapAchievementIncrementRequest {
            achievement_id: achievement_id_c.as_ptr(),
            steps,
        };

        let result = unsafe {
            tapsdk_pc_sys::TapAchievement_AsyncIncrement(self.handle, request_id, &request)
        };

        check_sdk_result(result)
    }

    /// Open the TapTap achievements page.
    pub fn show_achievements(&self) -> Result<()> {
        let result = unsafe { tapsdk_pc_sys::TapAchievement_ShowAchievements(self.handle) };

        check_sdk_result(result)
    }
}

fn check_sdk_result(result: u32) -> Result<()> {
    match SdkResult::from(result) {
        SdkResult::Ok => Ok(()),
        result => Err(TapSdkError::SdkRequestFailed(result)),
    }
}
